File: 0001~0500/0377. Combination Sum IV/main.py
"""
Source: https://leetcode.com/problems/combination-sum-iv/description/
Date: 2023/1/2
Skill: Dynamic Programming
Runtime: 40 ms, faster than 90.7%
Memory Usage: 13.9 MB, less than 79.87%
Time complexity:
Space complexity:
Constraints:
    1 <= nums.length <= 200
    1 <= nums[i] <= 1000
    All the elements of nums are unique.
"""

from typing import List

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        nums.sort()
        dp = [0 for _ in range(target+1)]
        dp[0] = 1

        for i in range(1, target + 1):
            for num in nums:
                if num > i: break
                dp[i] += dp[i-num]

        return dp[target]

if __name__ == "__main__":
    s = Solution

File: 0001~0500/0377. Combination Sum IV/main.cpp
/**
 * Source: https://rb.gy/f8dum
 * Date: 2023/9/9
 * Skill:
 * Runtime: 0 ms, faster than 100% of C++ online submissions
 * Memory Usage: 6.84 MB, less than 12.44% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>

using namespace std;

#define ll long long

typedef pair<int, int> pairs;


class Solution {
    vector<int> dp;
public:
    int combinationSum4(vector<int>& nums, int target) {
        dp.resize(target + 1, -1);
        return helper(nums, target);
    }
    int helper(vector<int>& nums, int target) {
        if (target == 0)
            return 1;
        if (dp[target] != -1)
            return dp[target];
        int res = 0;
        for (auto n: nums)
            if (n <= target)
                res += helper(nums, target - n);
        dp[target] = res;
        return dp[target];
    }
};


static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

int main() {
    Solution s;
    vector<int> nums{1,4,8,10,20};

}